src/internal/plugins/controllers/breadcrumbs.ts

```diff
@@ -0,0 +1,87 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+import debounce from '../../debounce';
+
+type ChangeCallback<T> = (props: T | null) => void;
+
+export interface BreadcrumbsGlobalRegistration<T> {
+  update(props: T): void;
+  cleanup(): void;
+}
+
+export interface BreadcrumbsApiInternal<T> {
+  registerAppLayout: (changeCallback: ChangeCallback<T>) => (() => void) | void;
+  registerBreadcrumbs: (props: T, onRegistered: () => void) => BreadcrumbsGlobalRegistration<T>;
+  getStateForTesting: () => {
+    appLayoutUpdateCallback: ChangeCallback<T> | null;
+    breadcrumbInstances: Array<{ props: T }>;
+    breadcrumbRegistrations: Array<() => void>;
+  };
+}
+
+export class BreadcrumbsController<T> {
+  #appLayoutUpdateCallback: ChangeCallback<T> | null = null;
+  #breadcrumbInstances: Array<{ props: T }> = [];
+  #breadcrumbRegistrations: Array<() => void> = [];
+
+  #notifyAppLayout = debounce(() => {
+    if (!this.#appLayoutUpdateCallback) {
+      return;
+    }
+    const latestBreadcrumb = this.#breadcrumbInstances[this.#breadcrumbInstances.length - 1];
+    this.#appLayoutUpdateCallback(latestBreadcrumb?.props ?? null);
+  }, 0);
+
+  #notifyBreadcrumbs = debounce(() => {
+    if (!this.#appLayoutUpdateCallback) {
+      return;
+    }
+    this.#breadcrumbRegistrations.forEach(listener => listener());
+  }, 0);
+
```

Comment:
We handle multiple breadcrumb instances, do we also need to handle multiple AppLayouts?

Reply:
Handled below. Only one app layout can register as breadcrumbs holder